app/services/discourse_reactions/reaction_manager.rb

```diff
@@ -7,27 +7,31 @@ def initialize(reaction_value:, user:, guardian:, post:)
       @user = user
       @guardian = guardian
       @post = post
+      @like = @post.post_actions.find_by(user: @user, post_action_type_id: post_action_like_type)
     end
 
     def toggle!
       ActiveRecord::Base.transaction do
+        return if (@like && !@guardian.can_delete_post_action?(@like)) || (is_reacted_by_user && !@guardian.can_delete_reaction_user?(old_reacted_user))
+        @reaction = reaction_scope&.first_or_create
+        @reaction_user = reaction_user_scope
         @reaction_value == DiscourseReactions::Reaction.main_reaction_id ? toggle_like : toggle_reaction
       end
     end
 
     private
 
     def toggle_like
-      like = @post.post_actions.find_by(user: @user, post_action_type_id: post_action_like_type)
-      raise Discourse::InvalidAccess if like && !@guardian.can_delete_post_action?(like)
-      like ? remove_shadow_like : add_shadow_like
+      remove_shadow_like if @like
+      remove_reaction if is_reacted_by_user
+      add_shadow_like unless @like
     end
 
     def toggle_reaction
       PostAction.limit_action!(@user, @post, post_action_like_type)
-      @reaction = reaction_scope.first_or_create
-      @reaction_user = reaction_user_scope&.first_or_initialize
-      @reaction_user.persisted? ? remove_reaction : add_reaction
+      remove_reaction if is_reacted_by_user
```

Comment:
I would do:

```ruby
if is_reacted_by_user
  remove_reaction
else
  add_reaction
end
```

Reply:
in this case, if the user wants to change its reaction, we need to first remove the reaction and then add a new one

Reply:
so you mean, we count on the fact that between both calls the result of is_reacted_by_user  will have changed?

Ok, well that's slightly error prone, but let's go with this for now

Reply:
I also got the same of changing is_reacted_by_user, added a fix in the latest commit